wgengine/magicsock/magicsock.go

```diff
@@ -2814,59 +2867,20 @@ func (c *Conn) ParseEndpoint(endpointStr string) (conn.Endpoint, error) {
 		return nil, fmt.Errorf("magicsock: ParseEndpoint: json.Unmarshal failed on %q: %w", endpointStr, err)
 	}
 	pk := key.Public(endpoints.PublicKey)
-	discoKey := endpoints.DiscoKey
 
 	c.mu.Lock()
 	defer c.mu.Unlock()
 	if c.closed {
 		return nil, errConnClosed
 	}
-	node, ok := c.peerMap.nodeForNodeKey(tailcfg.NodeKey(pk))
-	if !ok {
-		// We should never be telling WireGuard about a new peer
-		// before magicsock knows about it.
-		c.logf("[unexpected] magicsock: ParseEndpoint: unknown node key=%s", pk.ShortString())
-		return nil, fmt.Errorf("magicsock: ParseEndpoint: unknown peer %q", pk.ShortString())
-	}
-	de := &endpoint{
-		c:             c,
-		publicKey:     tailcfg.NodeKey(pk), // peer public key (for WireGuard + DERP)
-		wgEndpoint:    endpointStr,
-		sentPing:      map[stun.TxID]sentPing{},
-		endpointState: map[netaddr.IPPort]*endpointState{},
-	}
-	if !discoKey.IsZero() {
-		de.discoKey = tailcfg.DiscoKey(discoKey)
-		de.discoShort = de.discoKey.ShortString()
-	}
-	de.initFakeUDPAddr()
-	c.logf("magicsock: ParseEndpoint: key=%s: disco=%s; %v", pk.ShortString(), discoKey.ShortString(), logger.ArgWriter(func(w *bufio.Writer) {
-		const derpPrefix = "127.3.3.40:"
-		if strings.HasPrefix(node.DERP, derpPrefix) {
-			ipp, _ := netaddr.ParseIPPort(node.DERP)
-			regionID := int(ipp.Port())
-			code := c.derpRegionCodeLocked(regionID)
-			if code != "" {
-				code = "(" + code + ")"
-			}
-			fmt.Fprintf(w, "derp=%v%s ", regionID, code)
-		}
-
-		for _, a := range node.AllowedIPs {
-			if a.IsSingleIP() {
-				fmt.Fprintf(w, "aip=%v ", a.IP())
-			} else {
-				fmt.Fprintf(w, "aip=%v ", a)
-			}
-		}
-		for _, ep := range node.Endpoints {
-			fmt.Fprintf(w, "ep=%v ", ep)
-		}
-	}))
-	de.updateFromNode(node)
-	c.peerMap.upsertDiscoEndpoint(de)
+	if ep, ok := c.peerMap.endpointForNodeKey(tailcfg.NodeKey(pk)); ok {
+		return ep, nil
+	}
 
-	return de, nil
+	// We should never be telling WireGuard about a new peer
```

Comment:
Does the wireguard reconfig block until all calls to ParseEndpoint are done? If not, two rapid updates could add and remove an endpoint, with the slow following ParseEndpoint coming in after the second one and not finding the node key.

Reply:
Yes, all ParseEndpoint happens synchronously during Wireguard reconfig. Additionally, we hold the wgengine lock during the combination of the magicsock and wireguard reconfig, so we should be guaranteed that updates always happen in lockstep: first we tell magicsock about a new peer, then we tell wireguard about the same peer.